open-apparel-registry/open-apparel-registry
# Add sector to FacilityIndex

## Overview

Adds support for storing all `Facility` sectors on `FacilityIndex` as part of `index_facilities`

Connects #1831 


## Notes

This was fairly annoying to resolve, and I'm _not_ very happy about where I ended up.

This is the query that gets data for `FacilityIndex` right now:
```
SELECT          "api_facility"."name"            AS "name",
                "api_facility"."country_code"    AS "country_code",
                "api_facility"."location"::bytea AS "location",
                Array_agg("api_contributor"."contrib_type" ) filter (WHERE (
                                "api_facilitymatch"."is_active" = TRUE
                AND             "api_facilitymatch"."status" IN ('AUTOMATIC',
                                                                 'CONFIRMED',
                                                                 'MERGED')
                AND             "api_source"."is_active" = TRUE
                AND             "api_source"."is_public" = TRUE)) AS "contrib_types",
                array_agg("api_source"."contributor_id" ) filter (WHERE (
                                "api_facilitymatch"."is_active" = TRUE
                AND             "api_facilitymatch"."status" IN ('AUTOMATIC',
                                                                 'CONFIRMED',
                                                                 'MERGED')
                AND             "api_source"."is_active" = TRUE
                AND             "api_source"."is_public" = TRUE)) AS "contributors",
                array_agg("api_source"."facility_list_id" ) filter (WHERE (
                                "api_facilitymatch"."is_active" = TRUE
                AND             "api_facilitymatch"."status" IN ('AUTOMATIC',
                                                                 'CONFIRMED',
                                                                 'MERGED')
                AND             "api_source"."is_active" = TRUE
                AND             "api_source"."is_public" = TRUE))                                                                                                                                AS "lists",
                "api_facilitymatch"."facility_id"                                                                                                                                                AS "id",
                                concat("api_facility"."ppe_product_types", concat("api_facility"."ppe_contact_phone", concat("api_facility"."ppe_contact_email", "api_facility"."ppe_website"))) AS "ppe"
FROM            "api_facilitymatch"
inner join      "api_facility"
ON              (
                                "api_facilitymatch"."facility_id" = "api_facility"."id")
inner join      "api_facilitylistitem"
ON              (
                                "api_facilitymatch"."facility_list_item_id" = "api_facilitylistitem"."id")
inner join      "api_source"
ON              (
                                "api_facilitylistitem"."source_id" = "api_source"."id")
left outer join "api_contributor"
ON              (
                                "api_source"."contributor_id" = "api_contributor"."id")
GROUP BY        "api_facility"."name",
                "api_facility"."country_code",
                "api_facility"."location",
                "api_facilitymatch"."facility_id",
                                concat("api_facility"."ppe_product_types", concat("api_facility"."ppe_contact_phone", concat("api_facility"."ppe_contact_email", "api_facility"."ppe_website"))) 
```

We can't simply follow the same pattern as the scalar fields by using `ArrayAgg` as it would simply give us an array-of-arrays, but we also can't directly use `unnest` inside `ArrayAgg` either - this is apparently not allowed by Postgres. What Postgres wants us to do instead is the following:
```diff
@@ -22,6 +22,13 @@ SELECT          "api_facility"."name"            AS "name",
                                                                  'MERGED')
                 AND             "api_source"."is_active" = TRUE
                 AND             "api_source"."is_public" = TRUE))                                                                                                                                AS "lists",
+                array_agg(DISTINCT sectors) filter (WHERE (
+                                "api_facilitymatch"."is_active" = TRUE
+                AND             "api_facilitymatch"."status" IN ('AUTOMATIC',
+                                                                 'CONFIRMED',
+                                                                 'MERGED')
+                AND             "api_source"."is_active" = TRUE
+                AND             "api_source"."is_public" = TRUE))                                                                                                                                AS "sector",
                 "api_facilitymatch"."facility_id"                                                                                                                                                AS "id",
                                 concat("api_facility"."ppe_product_types", concat("api_facility"."ppe_contact_phone", concat("api_facility"."ppe_contact_email", "api_facility"."ppe_website"))) AS "ppe"
 FROM            "api_facilitymatch"
@@ -36,7 +43,8 @@ ON              (
                                 "api_facilitylistitem"."source_id" = "api_source"."id")
 left outer join "api_contributor"
 ON              (
-                                "api_source"."contributor_id" = "api_contributor"."id")
+                                "api_source"."contributor_id" = "api_contributor"."id")
+CROSS JOIN unnest("api_facilitylistitem"."sector") sectors
 GROUP BY        "api_facility"."name",
                 "api_facility"."country_code",
                 "api_facility"."location",
```

Unfortunately, this is not possible with the Django ORM - there is no built-in way to add an expression or subquery to the FROM-clause of the query.

I tried several different ways to rewrite the query as a subquery to work around that, but I couldn't come up with a good solution via that route either.

I considered making _just_ the `sector` portion it's own raw query, but given the complexity of the query that felt like potentially more of a maintenance burden than the solution I ended up with, even if it is a bit hacky.


 ---- 


I also fixed the `makecsvs` and `makefixtures` command to support `sector`,  in order to follow the instructions in https://github.com/open-apparel-registry/open-apparel-registry/pull/1214 on how to create large amounts of test data

## Testing Instructions

* `scripts/server`
* Add a CSV or post to the `facility-list` ViewSet. It should correctly update the `FacilityIndex` in your database

## Checklist
